src/cargo/util/errors.rs

```diff
@@ -300,6 +302,140 @@ impl From<clap::Error> for CliError {
     }
 }
 
+/// Error that comes from running `git` CLI. This wrapper exists to detect
+/// this kind of errors in [`crate::util::network::with_retry`] and properly
+/// retry them.
+pub struct GitCliError {
+    inner: anyhow::Error,
+}
+
+impl GitCliError {
+    /// Creates an insteance of [`GitCliError`].
+    ///
+    /// # Invariants
+    ///
+    /// The inner error must by constructed from [`ProcessError`] that
+    /// was returned as a result of running `git` CLI operation.
+    /// If it isn't true, the code won't panic, but [`Self::is_retryable()`]
+    /// will always return `false`.
```

Comment:
Instead of listing this as an invariant I think it might be better to do the downcast on construction of a `GitCliError` and have the internals stored as a `ProcessError` directly. This constructor could then return `Result<Self>` or continue to return `Self` and assert that the downcast succeeds.

Reply:
I am a bit scared to make this an `assert!()`, because I added only unit tests that don't invoke the code of `fetch_with_git_cli()`. I guess I should add a mocked binary implementation of `git` to verify this works more rigorously. Otherwise, I am afraid that this code is invoked very randomly and catching potential panics of cargo in the asserts is non-deterministic =(

Reply:
This API was just added here so I think it's fine to have an assert here. It's true, though, that `git-fetch-with-cli` is not heavily tested in Cargo's CI. If you'd like, though, then you could make the constructor argument here `ProcessError` and then when this is invoked you can use `error.downcast()?` to automatically propagate non-`ProcessError` errors and only pass in the process error here.

Reply:
Okay, thanks for the suggestions, I'll try to figure this out!